### ISSUE TITLE
PyVista next steps

### ISSUE THREAD

**Issue by banesullivan** [GeoStat-Framework/GSTools#32]
Now that #29 is wrapping up, what's next for 3D analysis and interoperability with VTK/PyVista for GSTools?

I think it would be awesome to get the following features:

- [x] ability to go from most VTK/PyVista mesh types to a GSTools field dataset
- [ ] Add a new `.plot` method that will use PyVista as the backend instead of Matplotlib for 3D datasets when PyVista is available.
- [ ] More (real-world) 3D examples!! With PyVista handling all the viz, we could easily showcase reproducible 3D examples which I haven't seen many of in the docs. Also, PyVista has many datasets in its `examples` module that could be used here (also see https://github.com/GeoStat-Framework/GSTools/issues/31#issuecomment-535258228)

And what about interoperabilty with other GSLib based software?
- PVGeo (extension of PyVista) has readers/writers for GSLib/SGeMS: https://pvgeo.org/examples/index.html#gslib
- Maybe PyVista could be a linking point between [MPSLib](https://mpslib.readthedocs.io/en/latest/) and GSTools?

**Comment by MuellerSeb** [comment]
Hey there!

I really like enthusiasm! :-D

It would be cool, to allow pyvista meshes as input for the Field.mesh method.
https://github.com/GeoStat-Framework/GSTools/blob/227478e25e77ad2ca57702f6fdbf10b5fdb6913e/gstools/field/base.py#L68
At the moment this method takes [ogs5py.MSH](https://github.com/GeoStat-Framework/ogs5py/blob/37c92c9a6d04e56603d001f323abc368bbb2cc10/ogs5py/fileclasses/msh/core.py#L1171) or [meshio.Mesh](https://github.com/nschloe/meshio/blob/caccb27aba6b634f00f3a8b26dbe84dbd5962e64/meshio/_mesh.py#L4) classes as input. Everything works implicit without importing these packages, so we don't have to check if these are installed or not.

I don't know if a new plot method is necessary at the moment. How does PyVista handle unstructured data points? Just as a cloud, or are ther interpolation options to show a convex hull?

More real world examples would be awesome since eye-catchy stuff is really most convincing for potential users.

What do you have in mind when talking about interoperabilty?

Cheers,
Sebastian

**Comment by MuellerSeb** [comment]
Also PyVista could be interessting to visualize the Output of [ogs5py](https://github.com/GeoStat-Framework/ogs5py):

https://github.com/GeoStat-Framework/ogs5py/blob/37c92c9a6d04e56603d001f323abc368bbb2cc10/ogs5py/fileclasses/msh/viewer.py#L20

https://github.com/GeoStat-Framework/ogs5py/blob/37c92c9a6d04e56603d001f323abc368bbb2cc10/ogs5py/tools/vtk_viewer.py#L16

Is PyVista capable of reading TecPlot and PVD files?

**Comment by banesullivan** [comment]
> Is PyVista capable of reading TecPlot and PVD files?

These formats could easily be added to PyVista's `read` routine. Do you have sample files for these formats?

> It would be cool, to allow pyvista meshes as input for the Field.mesh method.

Yes, definitely. This is likely low hanging fruit and could be implemented pretty easily.

> I don't know if a new plot method is necessary at the moment. How does PyVista handle unstructured data points? Just as a cloud, or are ther interpolation options to show a convex hull?

It can be either a point cloud or you could use an interpolation scheme like a 3D Delaunay filter to get a convex hull. PyVista's plotting is a lot more sophisticated than Matplotlib's so there's a lot more you can do like widgets for slicing through structured grids, thresholding volumes with widgets, etc. The best route forward is likely to just use the `to_pyvista` method and let users manually plot with PyVista.

> What do you have in mind when talking about interoperability?

Really just making sure going back and forth between PyVista mesh types and GeoStat mesh types is streamlined. A few plotting routines over here that call back to PyVista would also be nice.

**Comment by MuellerSeb** [comment]
#65 is related.

**Comment by MuellerSeb** [comment]
Just to take a note, when [vtk9](https://gitlab.kitware.com/vtk/vtk/-/milestones/15) is there and we have pre-built wheels for mac/linux/win for py35-py38 and pyvista is supporting vtk9, we can think about including pyvista as the plotting backend for fields.

To keep track: https://github.com/pyvista/pyvista/issues/562

@banesullivan, @LSchueler: what do you think?

**Comment by banesullivan** [comment]
I'd be excited to help implement a plotting backend with PyVista when the VTK9 wheels land on PyPI and we work out all the compatibility quirks with PyVista.

Once those wheels land, having PyVista as an optional dependency would be incredibly straightforward

**Comment by MuellerSeb** [comment]
@banesullivan : Since we are thinking about allowing arbitrary dimensions, does pyvista/vtk provide routines to deal with high-dimensional data? Like cutting out sections to 1D/2D/3D or projecting to lower dimensions?
For example, if we have a 4D field where we interpret the last dimension as time, it would be nice to have a 3D visualization with a slider for the time-dimension.
Same could be applied to 5D with two sliders and so on.

**Comment by banesullivan** [comment]
Interesting idea... we have the infrastructure to handle that kind of 3D rendering with slider bars and changing the shown data, however, PyVista does not have any straightforward way to hold time-varying or higher-dimensional data. You'd have to keep the data structure over here in GSTools and update the PyVista/VTK mesh on request. (This would be rather easy to implement if you have the data structures over here). The folks at [MNE-Python](https://github.com/mne-tools/mne-python) have been doing something very similar with PyVista for a while: see gif in https://github.com/mne-tools/mne-python/pull/7589#issue-401368766

**Comment by MuellerSeb** [comment]
Sth like [this](https://mne.tools/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#sphx-glr-auto-tutorials-source-modeling-plot-visualize-stc-py) would be cool.

Is it possible to have call-back functions for the sliders?

**Comment by GuillaumeFavelier** [comment]
Yes indeed! Through the `callback` parameter of `add_slider_widget()`. And many more options are available.

Reference: https://docs.pyvista.org/plotting/plotting.html?highlight=add_slider_widget#pyvista.BasePlotter.add_slider_widget

**Comment by MuellerSeb** [comment]
Since we now have interfaces for PyVista (back and forth), I count this as enabling plotting (as done in the documentation already).
Further 3D examples should live in https://github.com/GeoStat-Examples where we have more freedome to show what is possible!

Will close this one. Thanks for your support!
